pgmpy/factors/distributions/ContinuousDistribution.py

```diff
@@ -0,0 +1,230 @@
+
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> from scipy.special import beta
+        >>> from pgmpy.factors.distributions import CustomDistribution
+        # Two variable dirichlet distribution with alpha = (1, 2)
+        >>> def dirichlet_pdf(x, y):
+        ...     return (np.power(x, 1) * np.power(y, 2)) / beta(x, y)
+        >>> dirichlet_dist = CustomDistribution(variables=['x', 'y'],distribution=dirichlet_pdf)
+        >>> dirichlet_dist.scope()
+        ['x', 'y']
+        """
+        if not isinstance(variables, (list, tuple, np.ndarray)):
+            raise TypeError(
+                "variables: Expected type: iterable, got: {type}".format(
+                    type=type(variables)))
+
+        if len(set(variables)) != len(variables):
+            raise ValueError("Multiple variables can't have the same name")
+
+        self.variables = list(variables)
+        self._pdf = distribution
+
+    def reduce(self, values, inplace=True):
+        """
+        Reduces the factor to the context of the given variable values.
+
+        Parameters
+        ----------
+        values: list, array-like
+            A list of tuples of the form (variable_name, variable_value).
+
+        inplace: boolean
+            If inplace=True it will modify the factor itself, else would return
+            a new CustomDistribution object.
+
+        Returns
+        -------
+        CustomDistribution or None:
+                    if inplace=True (default) returns None
+                    if inplace=False returns a new ContinuousFactor instance.
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> from scipy.special import beta
+        >>> from pgmpy.factors.distributions import CustomDistribution
+        >>> def custom_pdf(x, y, z):
+        ...     return z*(np.power(x, 1) * np.power(y, 2)) / beta(x, y)
+        >>> custom_dist = CustomDistribution(['x', 'y', 'z'], custom_pdf)
+        >>> custom_dist.variables
+        ['x', 'y', 'z']
+        >>> custom_dist.assignment(1, 2, 3)
+        24.0
+
+        >>> custom_dist.reduce([('y', 2)])
+        >>> custom_dist.variables
+        ['x', 'z']
+        >>> custom_dist.assignment(1, 3)
+        24.0
+        """
+        if not isinstance(values, (list, tuple, np.ndarray)):
+            raise TypeError("variables: Expected type: iterable, "
+                            "got: {var_type}".format(var_type=type(values)))
+
+        for var, value in values:
+            if var not in self.variables:
+                raise ValueError("{var} not in scope.".format(var=var))
+
+        phi = self if inplace else self.copy()
+
+        var_to_remove = [var for var, value in values]
+        var_to_keep = [var for var in self.variables if var not in var_to_remove]
+
+        reduced_var_index = [(self.variables.index(var), value) for var, value in values]
+        pdf = self._pdf
+
+        def reduced_pdf(*args, **kwargs):
```

Comment:
`reduced_pdf` is a closure that captures the local variables in the `reduce` method. It is later assigned to `phi._pdf`, and becomes a private method of `phi`, which could be a different object than `self`. Do you think it better to have `self` and `phi` to share the same method?